### PR TITLE
MOBT-494: Model blending of spot forecasts

### PR DESCRIPTION
**Pull request by bayliffe**
Part of https://github.com/metoppv/mo-blue-team/issues/494

Changes required to enable model blending of spot forecasts. These are simply modifications to the way slicing is performed, removing an assumption of 2-dimensional spatial coordinates.

Acceptance test data: https://github.com/metoppv/improver_test_data/pull/10

Notebook demonstrating equivalence of model blending then spot-extraction, vs. spot-extraction then model blending: https://github.com/MetOffice/improver_aux/pull/372

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Review by brhooper — DISMISSED** [review]
Thanks @bayliffe, I can't see any issues here so I'm happy to approve this.

File: improver_tests/blending/weights/test_ChooseWeightsLinear.py [hunk 1]
```diff
@@ -328,43 +386,59 @@ def test_3d_array(self):
         self.assertArrayAlmostEqual(weights, expected_weights)
 
 
-class Test__create_new_weights_cube(IrisTest):
-    """Test the _create_new_weights_cube function. """
+"""Test the _create_new_weights_cube function. """
```
Comment (MoseleyS) [inline improver_tests/blending/weights/test_ChooseWeightsLinear.py:389]: This is no longer a doc-string and has become an active line with no purpose. Would this be better?
```suggestion
# Test the _create_new_weights_cube function
```

File: improver_tests/blending/weights/test_ChooseWeightsLinear.py [hunk 2]
```diff
@@ -328,43 +386,59 @@ def test_3d_array(self):
...
+@pytest.fixture
+def weights():
+    return np.array([0.0, 0.0, 0.2])
+
+
+@pytest.fixture(
+    params=[set_up_basic_model_config_cube, set_up_basic_model_config_spot_cube]
+)
+def single_thresh_input_cube(request, plugin):
+    return plugin._slice_input_cubes(request.param())[0]
```
Comment (MoseleyS) [inline improver_tests/blending/weights/test_ChooseWeightsLinear.py:406]: While these are really simple, I think each one should have a doc-string.

File: improver_tests/blending/weights/test_ChooseWeightsLinear.py [hunk 3]
```diff
@@ -328,43 +386,59 @@ def test_3d_array(self):
...
+    # test only relevant coordinates have been retained on the weights cube
+    expected_coords = {
+        "time",
+        "forecast_reference_time",
+        "forecast_period",
+        "model_id",
+        "model_configuration",
+    }
+    result_coords = {coord.name() for coord in new_weights_cube.coords()}
+    assert set(result_coords) == set(expected_coords)
```
Comment (MoseleyS) [inline improver_tests/blending/weights/test_ChooseWeightsLinear.py:427]: `result_coords` and `expected_coords` are already `set`s
```suggestion
    assert result_coords == expected_coords
```

File: improver_tests/blending/weights/test_ChooseWeightsLinear.py [hunk 4]
```diff
@@ -328,43 +386,59 @@ def test_3d_array(self):
...
+
+
+def test_new_weights_with_dict_masked_input(single_thresh_input_cube, weights, plugin):
+    """Test a new weights cube is created as intended when we have a masked
+    input gridded or spot forecast cube."""
+    single_thresh_input_cube.data = np.ma.masked_array(
+        single_thresh_input_cube.data,
+        np.ones(single_thresh_input_cube.data.shape) * True,
```
Comment (MoseleyS) [inline improver_tests/blending/weights/test_ChooseWeightsLinear.py:435]: There is no point in `* True` here as `True` takes the value of 1.

Please replace with one of:
`np.ones(single_thresh_input_cube.data.shape)`
or
`np.full(single_thresh_input_cube.data.shape, True)`

File: improver_tests/blending/weights/test_ChooseWeightsLinear.py [hunk 5]
```diff
@@ -420,57 +494,60 @@ def test_above_range_dict(self):
         self.assertEqual(new_weights_cube.name(), "weights")
 
 
-class Test__slice_input_cubes(IrisTest):
-    """Test the _slice_input_cubes method"""
+"""Test the _slice_input_cubes method"""
```
Comment (MoseleyS) [inline improver_tests/blending/weights/test_ChooseWeightsLinear.py:497]: Another orphaned doc-string
```suggestion
# Test the _slice_input_cubes method
```

File: improver_tests/blending/weights/test_ChooseWeightsLinear.py [hunk 6]
```diff
@@ -420,57 +494,60 @@ def test_above_range_dict(self):
...
-        self.assertEqual(len(result), 1)
-        self.assertArrayAlmostEqual(result[0].data, single_cube.data)
-        self.assertEqual(result[0].metadata, single_cube.metadata)
+def test__slice_input_slices(plugin, multi_model_inputs):
+    """Test function slices out extra dimensions to leave only the spatial
+    dimensions. Tested using a cube with and without a threshold coordinate."""
+
+    threshold_cube, no_threshold_cube, reference_cubelist = multi_model_inputs
+
+    for cube in [threshold_cube, no_threshold_cube]:
```
Comment (MoseleyS) [inline improver_tests/blending/weights/test_ChooseWeightsLinear.py:529]: This feels like it ought to be parametrized

**Review by MoseleyS — CHANGES_REQUESTED** [review]
I've spotted a few things that could be tweaked.

File: improver_tests/blending/weights/test_ChooseWeightsLinear.py [hunk 7]
```diff
@@ -93,6 +96,61 @@ def set_up_basic_model_config_cube(frt=None, time_points=None):
...
+        frt = dt(2017, 1, 10, 3, 0)
+    if time_points is None:
+        time_points = [
+            dt(2017, 1, 10, 9, 0),
+            dt(2017, 1, 10, 10, 0),
+            dt(2017, 1, 10, 11, 0),
+        ]
+
+    time_coords = construct_scalar_time_coords(frt, None, frt)
+    time_coords = [crd for crd, index in time_coords]
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/blending/weights/test_ChooseWeightsLinear.py:119]: Done

File: improver_tests/blending/weights/test_ChooseWeightsLinear.py (hunk 1 above)
Comment (bayliffe) [reply improver_tests/blending/weights/test_ChooseWeightsLinear.py:389]: Sure.

File: improver_tests/blending/weights/test_ChooseWeightsLinear.py (hunk 2 above)
Comment (bayliffe) [reply improver_tests/blending/weights/test_ChooseWeightsLinear.py:406]: Added

File: improver_tests/blending/weights/test_ChooseWeightsLinear.py (hunk 3 above)
Comment (bayliffe) [reply improver_tests/blending/weights/test_ChooseWeightsLinear.py:427]: Removed additional set calls.

File: improver_tests/blending/weights/test_ChooseWeightsLinear.py (hunk 4 above)
Comment (bayliffe) [reply improver_tests/blending/weights/test_ChooseWeightsLinear.py:435]: Done.

File: improver_tests/blending/weights/test_ChooseWeightsLinear.py (hunk 5 above)
Comment (bayliffe) [reply improver_tests/blending/weights/test_ChooseWeightsLinear.py:497]: Changed to # type.

File: improver_tests/blending/weights/test_ChooseWeightsLinear.py (hunk 6 above)
Comment (bayliffe) [reply improver_tests/blending/weights/test_ChooseWeightsLinear.py]: Done.

**Pull request merged**